### DISCUSSION TITLE
Next.js re-render components on route change

### DISCUSSION THREAD

**Discussion by nikita-ks** [vercel/next.js#31981] [Help]
### What version of Next.js are you using?

12.0.4

### What version of Node.js are you using?

14.17.0

### What browser are you using?

Chrome, Safari

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel

### Describe the Bug

Hi, its not possible to initialize for ex mobx store in _app using context or smth like that, because Next.js re-render components on route change. Or if u have layout and u need to use useEffect inside (ex in header component) and it is depend on some props or store useEffect will be called all time. its a pitty. Even u have a dump component in ur _app.tsx/jsx which does not have any props it will be re-rendered :((

i understand that props change and will be new render, but if u want to do somethings which depends on render u have to write ur code more complex

**_app.tsx**
```
function App({ Component, pageProps}: AppProps):ReactElement {
  return (
      <MainLayout>
        <Component {...rest} />
      </MainLayout>
  );
}

export default App;
```

**MainLayout.tsx**
```
export const MainLayout:NextComponentType = ({ children }) => {
  return (
    <>
      <Header title="some title" nav={[{ title: 'home', link: '/' }]} />
      <h1>tes</h1>
      <main className="main">
        {children}
      </main>
    </>
  );
};
```

**Header.tsx**
```
 export const Header:FC<HeaderProps> = ({ title, nav }) => {
  useEffect(() => {
    console.log(title, nav);
  }, [nav, title]);

  return (
    <header className="header">
      <div className="container">
        <div className="header__in">
          <div className="header_title fz_logo fw_bold">{title}</div>
          <nav className="header__nav"><NavLinks items={nav} /></nav>
        </div>
      </div>
    </header>
  );
}; 
```
this useEffect will be called all time when u will change route
```
useEffect(() => {
    console.log(title, nav);
  }, [nav, title]);
```

### Expected Behavior

-

### To Reproduce

just copy my code

**Comment by sicarius97** [comment]
@nikita-ks Could you talk a little bit more about what you are trying to accomplish? useEffect functions run anytime their dependencies change or the component mounts by design, so this isn't necessarily just a nextjs thing, but there are some ways to accomplish some of the things you mentioned. If you are saying you wouldn't want the example useEffect you supplied when changing routes, I am not sure that is the right place to use a useEffect hook anyway. If you changed pages, presumably the title and nav for said pages would also change, and being dependencies of that useEffect function, it would trigger the function no matter what because the state of the deps in that array changed

Components/pages in nextjs don't rerender everytime by default I believe, especially when shallow routing by using next/link

**Comment by nikhil-chougule** [comment]
I also faced a similar issue.
I am using the wrapper component to provide the context in _app -> here I wanted to extract query param and keep for that session.
Expected: 
  the context used to wrap the components in _app should be able to store extracted query params at first render.
Issue : 
  the whole app is re-rendering if I try to navigate after the first render causing the state to reset. afterwords it works fine.
I did a workaround. See if this helps 
```

    useEffect(()=>{
        if(!router.isReady) return;
       // add few additional checks
       // only if state is not set
       // only if expected query prams are present
       // now rerender the page manually 
       
       // set the  state with data then reload the page
       router.replace({
            pathname: router.pathname,
            query: router.query
        })
        // codes using router.query
    
    }, [router, router.isReady, router.query ]);
```

**Comment by JackDan9** [comment]
I also had the same problem
When I use Link to go to the next level of routing, re-render happens

**Comment by zhangwei900808** [comment]
same problem who can help me

**Comment by mehdiyahyavi** [comment]
I'm using antd components. have the same problem. when route change, all components re-render
Nextjs version: 13.4.4
React version: 18.2
AntD version: 5.5.2
